Python/pylifecycle.c

```diff
@@ -3464,6 +3472,43 @@ wait_for_thread_shutdown(PyThreadState *tstate)
     Py_DECREF(threading);
 }
 
+/* Wait for the interpreter's reference count to reach zero.
+   See PEP 788. */
+static void
+wait_for_interp_references(PyInterpreterState *interp)
+{
+    assert(interp != NULL);
+    struct _Py_finalizing_threads *finalizing = &interp->threads.finalizing;
+    _Py_atomic_store_int_release(&finalizing->shutting_down, 1);
+    PyMutex_Lock(&finalizing->mutex);
+    if (_Py_atomic_load_ssize_relaxed(&finalizing->countdown) == 0) {
+        // Nothing to do.
+        PyMutex_Unlock(&finalizing->mutex);
+        return;
+    }
+    PyMutex_Unlock(&finalizing->mutex);
+
+    PyTime_t wait_ns = 1000 * 1000; // 1 millisecond
+
+    while (true) {
+        if (PyEvent_WaitTimed(&finalizing->finished, wait_ns, 1)) {
```

Comment:
Would it be possible to implement exponential backoff?

```
PyTime_t wait_max = (PyTime_t)1000 * 1000 * 1000; // 1 second
...
wait_ns *= 2;
wait_ns = Py_MIN(wait_ns, wait_max);
```

I'm not sure what should be maximum. Maybe 100 ms?